skynet/haproxy_agent.py

```diff
@@ -0,0 +1,97 @@
+import asyncio
+import math
+
+from fastapi import FastAPI, Request
+from pydantic import BaseModel
+
+from skynet.env import whisper_max_connections
+
+from skynet.logs import get_logger
+from skynet.modules.monitoring import CONNECTIONS_METRIC, TRANSCRIBE_GRACEFUL_SHUTDOWN, TRANSCRIBE_STRESS_LEVEL_METRIC
+
+log = get_logger(__name__)
+
+# Bundles a REST API server used by the autoscaler with a TCP server that is queried by the HAProxy agent.
+# The REST API is used to change and query the state of the system. The HAProxy agent server is closely
+# bound to the REST API server as the latter will inform HAProxy if the system was set to drain mode.
+
+autoscaler_rest_app = FastAPI()
+TRANSCRIBE_GRACEFUL_SHUTDOWN.set(0)
+haproxy_state = 'ready'
+
+
+def get_haproxy_lb_percentage():
+    conns = CONNECTIONS_METRIC._value.get()
+    perc = int(math.floor(conns * 100 / whisper_max_connections))
+    inverted = 100 - perc
+    if inverted <= 0:
+        return 1
+    return int(inverted)
+
+
+async def handle_tcp_request(reader: asyncio.StreamReader, writer: asyncio.StreamWriter):
+    response = f'up {haproxy_state} {get_haproxy_lb_percentage()}%\n'
+    if writer is not None:
+        try:
+            writer.write(response.encode())
+        except Exception as ex:
+            log.warning('TCP exception during haproxy-agent check.', ex)
+        finally:
+            await writer.drain()
```

Comment:
If the TCP connection is broken, won't this fail?

Reply:
nope